### DISCUSSION TITLE
Externally Referenced Items

### DISCUSSION THREAD

**Discussion by poagester** [DomCR/ACadSharp#325] [Q&A]
I have been going through objects in the debugger and combing through code here to determine if I can easily find all XRef objects contained in the dwg. I am having no luck. I saw another post about the database being implemented at another time, but I was wondering if there was currently a way to do this without the database.

**Comment by DomCR** [comment]
Hi @poagester,

Here is a quick sample of how to get all the insert entities what are an XRef:

```C#
	CadDocument doc = DwgReader.Read(path, this.onNotification);
	foreach (Insert insert in doc.Entities.OfType<Insert>()) 
	{
		if(insert.Block.Flags.HasFlag(Blocks.BlockTypeFlags.XRef) 
			|| insert.Block.Flags.HasFlag(Blocks.BlockTypeFlags.XRef))
		{
			string xref = insert.Block.BlockEntity.XrefPath;
		}
	}
```

If you only need the `BlockRecord` you can use the same code but instead of the entities go through the `BlockRecords` table.

Let me know if this helps.

**Reply by poagester** [reply]
I think there might be a typo in the `if` statement here.

**Reply by DomCR** [reply]
yep, the second flag should be XRefOverlay which are the 2 options for an XRef, sorry about that.

**Reply by poagester** [reply]
I'm probably misunderstanding something. I did not yield any results from inserts or block records. I have a dwg that contains 3 jpg's. Are these considered inserts or externally referenced objects? I'm still trying to grasp the anatomy of a dwg.

**Reply by DomCR** [reply]
Images are currently not supported by ACadSharp, you should be able to see them as entities once they are implemented.

I'll create an issue linked to this discussion to add the feature.

**Reply by DomCR** [reply]
For more information about the data that is stored in a dxf or dwg check this link with the documentation:
https://help.autodesk.com/view/OARX/2021/ENU/?guid=GUID-3A2FF847-BE14-4AC5-9BD4-BD3DCAEF2281
